File: morale-assistant-backend/src/main/java/hu/herolds/projects/moraleassistant/util/synthesizer/MaryTTSSynthesizer.java
package hu.herolds.projects.moraleassistant.util.synthesizer;

import hu.herolds.projects.moraleassistant.exception.SynthesizeException;
import hu.herolds.projects.moraleassistant.model.enums.Language;
import hu.herolds.projects.moraleassistant.util.SoundFilesUtils;
import lombok.extern.slf4j.Slf4j;
import marytts.LocalMaryInterface;
import marytts.exceptions.SynthesisException;
import marytts.util.data.audio.MaryAudioUtils;
import org.springframework.stereotype.Component;

import javax.sound.sampled.AudioInputStream;
import java.io.IOException;
import java.nio.file.Path;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;

@Slf4j
@Component("MaryTTSSynthesizerUtil")
public class MaryTTSSynthesizer implements Synthesizer {
    private final LocalMaryInterface localMaryInterface;
    private final SoundFilesUtils soundFilesUtils;

    public MaryTTSSynthesizer(
            final LocalMaryInterface localMaryInterface,
            final SoundFilesUtils soundFilesUtils) {
        this.localMaryInterface = localMaryInterface;
        this.soundFilesUtils = soundFilesUtils;
    }

    public Path synthesize(final String text) throws SynthesizeException {
        // synthesize
        final AudioInputStream audio;
        try {
            audio = localMaryInterface.generateAudio(text);
        } catch (SynthesisException e) {
            throw new SynthesizeException(e);
        }

        // write to output
        double[] samples = MaryAudioUtils.getSamplesAsDoubleArray(audio);


        Path resolve = null;
        try {
            resolve = soundFilesUtils.getNextFilePath();
            log.info("File path to write: [{}]", resolve.toString());
            MaryAudioUtils.writeWavFile(samples, resolve.toString(), audio.getFormat());
            log.info("Output written to: [{}]", resolve.toString());
            return resolve;
        } catch (final IOException e) {
            log.error("Could not write to file: " + (resolve != null ? resolve.toString() : ""), e);
            throw new SynthesizeException(e);
        }

    }

    @Override
    public Set<Language> getSupportedLanguages() {
        return Stream.of(Language.EN).collect(Collectors.toSet());
    }

}

File: morale-assistant-backend-kt/src/main/kotlin/hu/herolds/projects/morale/service/synthetize/MaryTTSSynthesizer.kt
package hu.herolds.projects.morale.service.synthetize

import hu.herolds.projects.morale.domain.enums.Language
import hu.herolds.projects.morale.domain.enums.Language.EN
import hu.herolds.projects.morale.exception.SynthesizeException
import marytts.LocalMaryInterface
import marytts.util.data.audio.MaryAudioUtils
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import javax.sound.sampled.AudioInputStream
import kotlin.io.path.createTempFile
import kotlin.io.path.readBytes

@Component("MaryTTSSynthesizerUtil")
class MaryTTSSynthesizer(
    private val localMaryInterface: LocalMaryInterface,
): Synthesizer {
    private val log = LoggerFactory.getLogger(javaClass)

    override val supportedLanguages: Set<Language> = setOf(EN)

    override fun synthesize(text: String): ByteArray {
        try {
            log.info("Synthesizing text: [$text]")
            val audio: AudioInputStream = localMaryInterface.generateAudio(text)

            val samples = MaryAudioUtils.getSamplesAsDoubleArray(audio)

            val audioFilePath = createTempFile(prefix = "sound", suffix = ".wav").toAbsolutePath()

            MaryAudioUtils.writeWavFile(samples, audioFilePath.toString(), audio.format)
            log.info("Audio content written to file: [$audioFilePath]")
            return audioFilePath.toFile().apply {
                deleteOnExit()
            }.readBytes()
        } catch (ex: Exception) {
            log.error("Could not synthesize and save audio file.", ex)
            throw SynthesizeException(cause = ex)
        }
    }
}